### ISSUE TITLE
Allow scale and translation parameters to have units

### ISSUE THREAD

**Issue by jni** [napari/napari#1701]
## 🚀 Feature

Some popular unit frameworks include [pint](https://pint.readthedocs.io/en/stable/), [astropy.units](https://docs.astropy.org/en/stable/units/), and [unyt](https://unyt.readthedocs.io/en/stable/). For scientists I think it would be great to write something like `scale=np.array([1, 0.2, 0.2]) * u['µm']` (pint syntax) and have the rest of the napari UI take the units into account, e.g. when displaying coordinates, measuring distances, or displaying scale bars.

**Comment by kevinyamauchi** [comment]
May be of interest:[ `pint` integration with `xarray`](https://github.com/xarray-contrib/pint-xarray)

**Comment by sofroniewn** [comment]
Picking up in the discussion of units happening in https://github.com/napari/napari/pull/1720#issuecomment-706778646 and https://github.com/napari/napari/pull/1720#issuecomment-706779755

Do you @GenevieveBuckley and @Czaki have a preferred library or way of dealing with units in Python or experience with the packages that @jni suggests?

If we get a PR adding unit support in first we can then extend that to #1720 to make sure the scale bar works properly

**Comment by Czaki** [comment]
I use custom implementation of units. But presented libraries looks great and I think that I should switch to one of them.

**Comment by GenevieveBuckley** [comment]
I don't have a lot of direct experience with this, but if pint can integrate with xarray I think that's a point in it's favour. 

I would like to see stronger xarray support here too, we might find that we get some of what we need here for free, or in a much easier way.

**Comment by sofroniewn** [comment]
> I don't have a lot of direct experience with this, but if pint can integrate with xarray I think that's a point in it's favour.

Agreed! Maybe @Czaki you can investigate switching over your work to pint and report back on if that looks like a good idea for us to adopt here.

@GenevieveBuckley let us know if you look in pint more too.

It seems relatively widely used at this point > 2000 projects, which is nice!!

**Comment by jni** [comment]
I've used pint before in small projects and it's great, others were included for completeness. Unyt is the new hotness and claims to compute things up to 2x as fast as pint (if I remember correctly), but I think for the uses we are likely to have in napari this is not a huge concern. It's also a much heavier dependency as it pulls down SymPy. So I'm pro-pint.

**Comment by Czaki** [comment]
I start to do a review of `unyt` and `pint`. 
The big disadvantage of `pint` is a lack of the global register of units:

> Warning

> There are no global units in Pint. All units belong to a registry and you can have multiple registries instantiated at the same time. However, you are not supposed to operate between quantities that belong to different registries. Never do things like this: ...

https://pint.readthedocs.io/en/stable/tutorial.html#using-pint-in-your-projects

This creates a big problem with integrating different codes. Especially when the plugin is the only wrapper of some package. Especially the default unit register needs to be defined outside the napari package (for example in `napari-plugin-engine`) to provide light dependencies to other packages. 

For example, I try to keep minimal dependencies on the core IO part of my tool to simply extract it to a separate package. And will be really unhappy if there is try to force to depend on thole `napari`.

Another example is some package that performs some measurements on data. And where napari allows creating widget plugin then would like to add an option to call its method from napari GUI. 

`unyt` package has `sympy` dependencies but has a global register of units. This simplifies the interaction between packages. Also `unyt` have builtin integration with `pint`. 

At this moment I prefer a little more `unyt`. But it is based only on first impressions. 

Did you have any questions/remarks connected with unit registers?

**Comment by sofroniewn** [comment]
Ok, good to know @Czaki, thanks!!!

> For example, I try to keep minimal dependencies on the core IO part of my tool to simply extract it to a separate package. And will be really unhappy if there is try to force to depend on thole napari.

Yeah, we definitely won't go down this road. I do wonder if there is some way around this though. I'll have to look in more.

I found some discussion here https://github.com/hgrecco/pint/issues/1071#issuecomment-610978025 and here https://github.com/xarray-contrib/pint-xarray/issues/7 that might be helpful. Not sure if @jni has experience

**Comment by jni** [comment]
Thanks for bringing up this issue @Czaki! Absolutely we do not want people to have to depend on napari in order to interoperate with our units. 100% on board there.

@sofroniewn I don't have experience with application registries (I've only ever used my own local one), but the links you provided seem to provide a solution for this? I'm still a little unclear how they work, but to me it seems there's still solutions for using pint? My hope is that we can make this work even without an external "napari-units" package or whatever.

**Comment by GenevieveBuckley** [comment]
@Czaki if napari ends up supporting more xarray data, do you think `unyt` will also work well in that situation? Thanks for investigating!

**Comment by Czaki** [comment]
@GenevieveBuckley  `unyt` units have a method `to_pint` and `from_pint` methods which allow simple integration with `pint`. I have no experience with `xarray` (I think that I should change it)

https://unyt.readthedocs.io/en/stable/usage.html#working-with-code-that-uses-pint

@jni As I good understand the solution is a global register and set it as default argument in wraped function.

**Comment by will-moore** [comment]
I'm looking for a way to pass units to napari when [opening an image via a plugin](https://github.com/ome/napari-ome-zarr/pull/31) and it looks like this issue is it.

So the plugin could return?
```
(data, {"scale": ([1, 0.2, 0.2]) * u['µm'])})
```
Thanks!

**Comment by sofroniewn** [comment]
I think we never actually added unit support here, but maybe now is a good time to pick this up again!

**Comment by sofroniewn** [comment]
cc @andy-sweet and @tlambert03 who might be interested in this

**Comment by Czaki** [comment]
I think that this is solved by `layer.units`

**Comment by TimMonko** [comment]
I agree that this has been solved by #6780 -- I am going to close now and we can re-open if needed.
